src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6455,10 +6455,19 @@ static NullableFlowState getReferenceConversionResultState(TypeWithAnnotations t
                 switch (state)
                 {
                     case NullableFlowState.MaybeNull:
-                        if (operandType.Type?.IsTypeParameterDisallowingAnnotationInCSharp8() != true &&
-                            targetType.Type?.IsTypeParameterDisallowingAnnotationInCSharp8() == true)
+                        if (targetType.Type?.IsTypeParameterDisallowingAnnotationInCSharp8() == true)
                         {
-                            return NullableFlowState.MaybeDefault;
+                            var type = operandType.Type;
+                            if (type?.IsTypeParameterDisallowingAnnotationInCSharp8() != true)
```

Comment:
nit: consider aligning with next local function `if (type is null || !type.IsTypeParameterDisallowingAnnotationInCSharp8())`